pkg/statement/statement.go

```diff
@@ -227,20 +228,48 @@ func (a *AbstractStatement) AlterContainsAddUnique() error {
 }
 
 // AlterContainsIndexVisibility checks to see if there are any clauses of an ALTER to change index visibility.
-// It really does not make sense for visibility changes to be anything except metadata only changes,
-// because they are used for experiments. An experiment is not rebuilding the table. If you are experimenting
-// setting an index to invisible and plan to switch it back to visible quickly if required, going through
-// a full table rebuild does not make sense.
+// It now allows index visibility changes when mixed with other metadata-only operations,
+// but blocks them when mixed with table-rebuilding operations to avoid semantic issues.
 func (a *AbstractStatement) AlterContainsIndexVisibility() error {
 	alterStmt, ok := (*a.StmtNode).(*ast.AlterTableStmt)
 	if !ok {
 		return ErrNotAlterTable
 	}
+
+	hasIndexVisibility := false
+	hasNonMetadataOperation := false
+
 	for _, spec := range alterStmt.Specs {
-		if spec.Tp == ast.AlterTableIndexInvisible {
-			return errors.New("the ALTER operation contains a change to index visibility and could not be completed as a meta-data only operation. This is a safety check! Please split the ALTER statement into separate statements for changing the invisible index and other operations")
+		switch spec.Tp {
+		case ast.AlterTableIndexInvisible:
+			hasIndexVisibility = true
+		case ast.AlterTableDropIndex,
+			ast.AlterTableRenameIndex,
+			ast.AlterTableDropPartition,
+			ast.AlterTableTruncatePartition,
+			ast.AlterTableAddPartitions,
+			ast.AlterTableAlterColumn:
+			// These are metadata-only operations - safe to mix with index visibility
+			continue
+		case ast.AlterTableModifyColumn, ast.AlterTableChangeColumn:
+			// Only safe if changing length of a VARCHAR column
+			if spec.NewColumns[0].Tp != nil && spec.NewColumns[0].Tp.GetType() == mysql.TypeVarchar {
+				continue
+			}
+			hasNonMetadataOperation = true
+		case ast.AlterTableAddConstraint: // ADD INDEX operations are table-rebuilding
+			hasNonMetadataOperation = true
+		default:
+			// All other operations are considered non-metadata (table rebuilding)
+			hasNonMetadataOperation = true
 		}
 	}
+
+	// Only fail if index visibility is mixed with non-metadata operations
+	if hasIndexVisibility && hasNonMetadataOperation {
+		return errors.New("the ALTER operation contains a change to index visibility mixed with table-rebuilding operations. This creates semantic issues for experiments. Please split the ALTER statement into separate statements for changing the invisible index and other operations")
+	}
+
```

Comment:
This semantic change makes sense to me. I wonder if we can move it into the checks/ system though so we keep the flow of a migration simple.

We can handle it as a followup though.